src/tfs/tlog.c

```diff
@@ -365,60 +561,73 @@ static void log_set_dirty(log tl)
 }
 #endif
 
-void log_write_eav(log tl, tuple e, symbol a, value v)
+boolean log_write_eav(log tl, tuple e, symbol a, value v)
 {
     tlog_debug("log_write_eav: tl %p, e %p, a %b, v %p\n", tl, e, symbol_string(a), v);
     u64 len = buffer_length(tl->tuple_staging);
-    encode_eav(tl->tuple_staging, tl->dictionary, e, a, v);
+    if (len >= bytes_from_sectors(tl->fs, range_span(tl->current->sectors)))
+        return false;
```

Comment:
It's not clear to me why this would fail (same for log_write). tl->tuple_staging can store up an arbitrary amount of encoded tuple data which can span multiple extensions; it should not be limited to the size of the current extension. A very large tuple can even span an extension boundary (this is why we have TUPLE_EXTENDED).

The only problem with this scheme is that an error on log extension creation (out of storage) would happen asynchronously, as a result of a log flush. But, much like reporting write errors in tfs, perhaps it would suffice to record such an error in the log and report it on a subsequent log_write_* (imprecise error reporting).

Reply:
Both log_write() and log_write_eav() call log_set_dirty(), which has now been modified in order to call log_flush() directly if the staging buffer is "full enough". Because of that, the staging buffer (which is consumed when log_flush() is called) should never grow more than the "full enough" threshold used in log_set_dirty(); and log_write() and log_write_eav() check against a threshold double that in log_set_dirty(). So an overly full buffer indicates that a previous call to log_flush() was not successful.
But if you prefer I can replace those checks in log_write() and log_write_eav() with the error reporting scheme you mentioned (we would need to add an error flag in the tlog structure, which will be set if a log_flush cannot complete successfully, and log_write() and log_write_eav() would return false if this error flag is set).